contracts/EntryPoint.sol

```diff
@@ -319,34 +311,38 @@ contract EntryPoint is StakeManager {
     // create the sender's contract if needed.
     function _createSenderIfNeeded(MemoryUserOp memory mUserOp, bytes calldata initCode) internal {
         if (initCode.length != 0) {
-            // note that we're still under the gas limit of validate, so probably
-            // this create2 creates a proxy account.
-            // @dev initCode must be unique (e.g. contains the signer address), to make sure
-            //   it can only be executed from the entryPoint, and called with its initialization code (callData)
-            address sender1 = ICreate2Deployer(create2factory).deploy(initCode, bytes32(mUserOp.nonce));
-            require(sender1 != address(0), "create2 failed");
-            require(sender1 == mUserOp.sender, "sender doesn't match create2 address");
+            require(mUserOp.sender.code.length == 0, "sender already constructed");
+            address sender1 = _createSender(initCode);
+            require(sender1 == mUserOp.sender, "sender doesn't match initCode address");
+            require(sender1.code.length != 0, "initCode failed to create sender");
         }
     }
 
     /**
-     * Get counterfactual sender address.
-     *  Calculate the sender contract address that will be generated by the initCode and salt in the UserOperation.
-     * @param initCode the constructor code to be passed into the UserOperation.
-     * @param salt the salt parameter, to be passed as "nonce" in the UserOperation.
+     * call the "initCode" factory to create and return the sender wallet address
+     * initCode must be unique (e.g. contains the signer address), to make sure
+     *  it can only be executed from the entryPoint, and called with its initialization code (callData)
+     * @param initCode the initCode value from a UserOp. contains 20 bytes of factory address, followed by calldata
+     * @return sender the returned address of the created wallet.
      */
-    function getSenderAddress(bytes memory initCode, uint256 salt) public view returns (address) {
-        bytes32 hash = keccak256(
-            abi.encodePacked(
-                bytes1(0xff),
-                address(create2factory),
-                salt,
-                keccak256(initCode)
-            )
-        );
+    function _createSender(bytes calldata initCode) internal returns (address sender) {
+        address initAddress = address(bytes20(initCode[0 : 20]));
+        bytes memory initCallData = initCode[20 :];
+        bool success;
+        assembly {
+            success := call(gas(), initAddress, 0, add(initCallData, 0x20), mload(initCallData), 0, 32)
```

Comment:
Calling a user-provided address address from EntryPoint (which isn't the sender) seems risky.  Wallets trust EntryPoint to only call their execution functions after they approve it in validation.  Could an attacker abuse this call to perform unauthorized operations on 3rd party wallets?

Maybe something along these lines:
* Attacker creates a batch meant for a wallet that has internal batching.  The batch contains two operations.  One is a transfer of all the wallet's assets, and the other is a call to a factory contract that will create the expected `sender` (which doesn't exist yet).
* Attacker sends a UserOp with this non-existent `sender` and with an initcode where the first 20 bytes are the victim wallet's address and the rest of the initcode is the batch above.
* `_createSenderIfNeeded` gets called, verifies that `sender` has no code yet, and calls `_createSender`.
* `_createSender` calls the victim wallet, giving it the malicious batch with the two operations.
* The wallet accepts the call because it trusts EntryPoint.  It transfers all of its assets (or even changes its owner), and then also calls the factory to generate `sender`.
* When the wallet returns, `sender` has code and the creation is considered successful.
* `sender`'s validation is called, and since it's an attacker-controlled contract, it of course succeeds.
* The rest of the op is delivered to `sender` but it doesn't matter at this point.  The attack was already successful.

I haven't checked whether this particular flow will work, but this call seems risky.  Maybe we should move it to a separate helper contract that gets deployed from EntryPoint's constructor, so that the call doesn't come directly from EntryPoint.